### PR TITLE
Charlesmchen/permissions vs inactive

### PR DESCRIPTION
**Pull request by charlesmchen**
PTAL @michaelkirk

### CODE REVIEW

File: Signal/src/ViewControllers/ConversationView/ConversationViewController.m [hunk 1]
```diff
@@ -3047,27 +3050,25 @@ - (void)requestRecordingVoiceMemo
     self.voiceMessageUUID = voiceMessageUUID;
 
     __weak typeof(self) weakSelf = self;
-    [[AVAudioSession sharedInstance] requestRecordPermission:^(BOOL granted) {
```
Comment (charlesmchen) [inline Signal/src/ViewControllers/ConversationView/ConversationViewController.m]: Centralize the asking for camera and microphone permissions.

File: Signal/src/call/CallService.swift [hunk 2]
```diff
@@ -1103,33 +1103,34 @@ protocol CallServiceObserver: class {
...
-
-        // We don't need to worry about the user granting or remoting this permission
-        // during a call while the app is in the background, because changing this
-        // permission kills the app.
-        if authStatus != .authorized {
-
-            OWSAlerts.showAlert(withTitle:NSLocalizedString("MISSING_CAMERA_PERMISSION_TITLE", comment: "Alert title when camera is not authorized"),
-                                message:NSLocalizedString("MISSING_CAMERA_PERMISSION_MESSAGE", comment: "Alert body when camera is not authorized"))
-
+        guard let frontmostViewController = UIApplication.shared.frontmostViewController else {
```
Comment (charlesmchen) [inline Signal/src/call/CallService.swift]: Never ask for camera or microphone permissions if app is not active.  I think I'm going to modify this to only skip if in background.  iOS could temporarily make app inactive to show a system alert, e.g. "iOS update available."

File: Signal/src/util/UIViewController+Permissions.m [hunk 3]
```diff
@@ -0,0 +1,101 @@
...
+    DDLogVerbose(@"[%@] ows_askForCameraPermissions", NSStringFromClass(self.class));
+
+    // Ensure callback is invoked on main thread.
+    void (^callback)(BOOL) = ^(BOOL granted) {
+        DispatchMainThreadSafe(^{
+            callbackParam(granted);
+        });
+    };
+
+    if ([UIApplication sharedApplication].applicationState != UIApplicationStateActive) {
```
Comment (charlesmchen) [inline Signal/src/util/UIViewController+Permissions.m]: Never ask for camera or microphone permissions if app is not active. I think I'm going to modify this to only skip if in background. iOS could temporarily make app inactive to show a system alert, e.g. "iOS update available."

File: Signal/src/util/UIViewController+Permissions.m [hunk 4]
```diff
@@ -0,0 +1,101 @@
...
+
+@implementation UIViewController (Permissions)
+
+- (void)ows_askForCameraPermissions:(void (^)(BOOL granted))callbackParam
+{
+    DDLogVerbose(@"[%@] ows_askForCameraPermissions", NSStringFromClass(self.class));
+
+    // Ensure callback is invoked on main thread.
+    void (^callback)(BOOL) = ^(BOOL granted) {
+        DispatchMainThreadSafe(^{
```
Comment (charlesmchen) [inline Signal/src/util/UIViewController+Permissions.m]: Make sure permissions callbacks are always invoked on main thread.

File: Signal/src/call/CallService.swift [hunk 5]
```diff
@@ -1434,7 +1435,7 @@ protocol CallServiceObserver: class {
         AssertIsOnMainThread()
 
         if case CallError.assertionError(description:let description) = error {
-            owsFail(description)
```
Comment (michaelkirk) [inline Signal/src/call/CallService.swift]: I don't think we should get rid of this. Is there some assertion that you're regularly seeing which is making this inconvenient? Maybe we should revisit that assertion.

File: Signal/src/util/UIViewController+Permissions.m [hunk 6]
```diff
@@ -0,0 +1,101 @@
...
+                                                                  callback(NO);
+                                                              }];
+        [alert addAction:dismissAction];
+
+        [self presentViewController:alert animated:YES completion:nil];
+    } else if (status == AVAuthorizationStatusAuthorized) {
+        callback(YES);
+    } else if (status == AVAuthorizationStatusNotDetermined) {
+        [AVCaptureDevice requestAccessForMediaType:AVMediaTypeVideo
+                                 completionHandler:^(BOOL granted) {
```
Comment (michaelkirk) [inline Signal/src/util/UIViewController+Permissions.m]: nit: unnecessary wrap: `completionHandler:callback`

File: Signal/src/util/UIViewController+Permissions.m [hunk 7]
```diff
@@ -0,0 +1,101 @@
...
+        });
+    };
+
+    if ([UIApplication sharedApplication].applicationState == UIApplicationStateBackground) {
+        DDLogError(@"Skipping microphone permissions request when app is in background.");
+        callback(NO);
+        return;
+    }
+
+    [[AVAudioSession sharedInstance] requestRecordPermission:^(BOOL granted) {
```
Comment (michaelkirk) [inline Signal/src/util/UIViewController+Permissions.m]: nit: unnecessary wrap.

Comment (charlesmchen) [reply Signal/src/util/UIViewController+Permissions.m]: Done.

File: Signal/src/util/UIViewController+Permissions.m (hunk 6 above)
Comment (charlesmchen) [reply Signal/src/util/UIViewController+Permissions.m]: Done.

File: Signal/src/call/CallService.swift (hunk 5 above)
Comment (charlesmchen) [reply Signal/src/call/CallService.swift]: I hit this assert while hanging up a test-device-to-simulator call.  I assumed it was yet another over-aggressive assert in `CallService`.  I'll restore it, but I'm not sure we should `owsFail` edge cases which will recover gracefully in production.
